fix mem leaks in try_decode_PKCS12() of crypto/store/

DDvO opened the pull request:
When loading PKCS#12 files that include a cert/ca chain the chain container was not deallocated.
Fixing this I improved the layout of `crypto/store/loader_file.c` to satisfy `check-format.pl -l`.

DDvO commented:
BTW, the mem leak came up experimenting how to use the OSSL_STORE API as requested by @mattcaswell  for the CMP contribution chunk 11: https://github.com/openssl/openssl/pull/11470#discussion_r416669564

It indicates that the OSSL_STORE API is not sufficiently tested. In fact, I can't find any tests for it.

mattcaswell commented on crypto/store/loader_file.c:
```diff
@@ -272,6 +272,7 @@ static OSSL_STORE_INFO *try_decode_PKCS12(const char *pem_name,
                         osi_ca = NULL;
                         (void)sk_X509_shift(chain);
                     }
+                    sk_X509_free(chain);
```
Shouldn't this "free" happen after the following "}" not before?

DDvO replied:
Right, the fix was not sufficient.
Having an even closer look I found a couple of more potential memory leaks in case of allocation errors. Moreover the function did not cope with cases that the pkey and/or cert in the PKCS#12 structure is not present.
I've fixed all these issues.

mattcaswell commented:
> It indicates that the OSSL_STORE API is not sufficiently tested. In fact, I can't find any tests for it.

There are lots of store tests in `90-test_store.t` - but they all test via the command line store app rather than the API directly.

DDvO commented:
> > It indicates that the OSSL_STORE API is not sufficiently tested. In fact, I can't find any tests for it.
> 
> There are lots of store tests in `90-test_store.t` - but they all test via the command line store app rather than the API directly.

Ah, right.
I've made sure that these tests can run also if DSA and/or EC key generation does not work
(BTW, I had to disable tests based on `pbeWithMD5AndDES-CBC` apparently due to deprecation)
and added `-chain -CAfile cacert.pem` to the `openssl pkcs12 -export` calls
such that memory leaks are caught also in case they are due to a non-empty chain.

DDvO commented:
Currently I always see one Travis CI failure like this:
```
 No space left on device at ./Configure line 1722.
```

DDvO commented:
>  No space left on device at ./Configure line 1722.

Gosh, my mem leak fixes did not solve this ;-)

levitte commented on crypto/store/loader_file.c:
```diff
@@ -14,16 +14,16 @@
-#include <openssl/pkcs12.h>      /* For the PKCS8 stuff o.O */
-#include <openssl/rsa.h>         /* For d2i_RSAPrivateKey */
+#include <openssl/pkcs12.h> /* For the PKCS8 stuff o.O */
+#include <openssl/rsa.h> /* For d2i_RSAPrivateKey */
```
Why?

levitte replied:
If this was for check-format, then I think check-format is asking too much.  In the original source-format stuff, we actually have column 33 as the starting column for these kinds of same-line one-line comments.

DDvO replied:
Well, as I wrote this is just to satisfy the checks of `check-format.pl`, which otherwise report "double space" at least if there is just one such comment not directly following code.

Actually for the two lines highlighted above the tool would not have reported this because the comments in two consecutive lines started on the same column.
Shall I revert this?

DDvO replied:
Ah, I just got your second comment.
So I've changed back the for first line to comments starting at column 33.

mattcaswell commented on test/recipes/90-test_store.t:
```diff
@@ -44,21 +44,32 @@ my @generated_files =
      "rsa-key-pkcs8-pbes2-sha256.pem", "rsa-key-pkcs8-pbes2-sha256.der",
-     "rsa-key-pkcs8-pbes1-md5-des.pem", "rsa-key-pkcs8-pbes1-md5-des.der",
+    );
+my $use_md5 = 0; # does not work: !disabled("md5");
```
Why does `!disabled("md5")` not work?

DDvO replied:
For some reason the "no-md5" config option is not (anymore?) usable.

Moreover, even if md5 is available calls like
```
../../../apps/openssl pkcs12 -inkey rsa-key-pkcs8.pem -in rsa-cert.pem
 -passout 'pass:password' -chain -CAfile cacert.pem -export -macalg SHA1
 -certpbe pbeWithMD5AndDES-CBC
 -keypbe pbeWithMD5AndDES-CBC -out rsa-key-md5-des-sha1.p12
```
may fail apparently because pbeWithMD5AndDES-CBC is deprecated.
How can one test at this point if pbeWithMD5AndDES-CBC is enabled?

DDvO replied:
Again my question: how to check in Perl whether pbeWithMD5AndDES-CBC is enabled?

levitte replied:
`pbeWithMD5AndDES-CBC` per se isn't deprecated as far as I can tell from finding it in the source.  However, DES-CBC is in the legacy provider, so you need to ensure it gets loaded (you want to use the config file `test/default-and-legacy.cnf` to have that).  Additionally, `des` may be disabled, sso if you want to check anything, it should be that.

DDvO replied:
Interesting hint. Since the pkcs12 app does not have a `-config` option I tried this way:
```
OPENSSL_CONF=test/default-and-legacy.cnf V=1 make  test TESTS=test_store | grep pbeWithMD5AndDES-CBC
```
but I still get
```
../../../util/wrap.pl ../../../apps/openssl pkcs12 -inkey rsa-key-pkcs8.pem -in rsa-cert.pem -passout 'pass:password' -chain -CAfile cacert.pem -export -macalg SHA1 -certpbe pbeWithMD5AndDES-CBC -keypbe pbeWithMD5AndDES-CBC -out rsa-key-md5-des-sha1.p12 => 1
```

levitte replied:
In perl, you would set it like this:

``` perl
$ENV{OPENSSL_CONF} = srctop_file("test", "default-and-legacy.cnf");
```

DDvO replied:
> In the perl recipe, you would set it like this:
> 
> ```perl
> $ENV{OPENSSL_CONF} = srctop_file("test", "default-and-legacy.cnf");
> ```

Thanks for the hint. 
Yet it took me a while to realize that, for whatever reason, the configuration file is ignored in the tests, but calling `openssl pkcs12 -provider default -provider legacy` does the trick.

levitte replied:
No, `OPENSSL_CONF` isn't ignored, it's *overwritten* by the test runner:

https://github.com/openssl/openssl/blob/00da0f69890874feaa555fafb99b967b861e9118/test/run_tests.pl#L30-L34

mattcaswell commented on test/recipes/90-test_store.t:
```diff
@@ -44,21 +44,32 @@ my @generated_files =
+    "rsa-key-md5-des-sha1.p12",
+    "rsa-key-aes256-cbc-md5-des-sha256.p12",
+    "rsa-key-pkcs8-pbes1-md5-des.pem", "rsa-key-pkcs8-pbes1-md5-des.der"
+     )) if $use_md5;
```
At the moment `$use_md5` is always 0...so this is redundant code.

DDvO replied:
I've tentatively changed its initialization to
```
my $use_md5 = !disabled("md5");
```
but see the issue with pbeWithMD5AndDES mentioned above.

mattcaswell commented on test/recipes/90-test_store.t:
```diff
@@ -44,21 +44,32 @@ my @generated_files =
+    "rsa-key-pkcs8-pbes1-md5-des.pem", "rsa-key-pkcs8-pbes1-md5-des.der"
+     )) if $use_md5;
+my $use_dsa = !disabled("dsa")
+              && run(app(["openssl", "dsaparam", "-genkey", "1024"]));
```
Why is running dsaparam here necessary?

DDvO replied:
This is because I found that even when DSA is enabled the key generation may fail
(maybe because 1024 is not any more considered secure enough).

DDvO replied:
Is there a better way to test this at this point?

DDvO replied:
Again my question: how to check with Perl if DSA-1024 is (not) deprecated?

t8m replied:
What if DSA-2048 was used instead?

DDvO replied:
> What if DSA-2048 was used instead?

Good thought. I've just tried it out but this did not help.

Yet since fixing the applicability of those test is just a side issue that I tried to tackle along with fixing the mem leaks in `try_decode_PKCS12()` (and the docs) I suggest leaving this topic for now, also because it's not urgent and there is already an approval.

levitte replied:
All this discussion doesn't explain why you have that run call here.  Isn't the one in the `init` function enough?

DDvO replied:
> > When loading PKCS#12 files that include a cert/ca chain the chain container was not deallocated.
> 
> Looking back, I'm surprised we haven't seen that already. Is there a way to demonstrate this with existing master, or even 1.1.1?

With the current master it's hard to reproduce because my commit 3c38fa4b797848a76b83f91e423de470adbb3b61, which is already merged, includes a preliminary fix, while this PR also fixes the mem leaks on malloc failures.

Moreover, the pkcs12 app does not use the STORE API for loading PKCS#12 files, and in 1.1.1 apparently the only app using it is `storeutl`.
And so far the tests did not include a chain in the PKCS#12 file(s) parsed.

Here is a way to demonstrate the mem leak for the 1.1.1 master if mdebug is enabled:
```
apps/openssl pkcs12 -inkey test/certs/interCA.key -in test/certs/interCA.pem -chain -CAfile test/certs/rootCA.pem -export -descert --passout pass: -out test.p12
apps/openssl storeutl test.p12
```

DDvO replied:
> All this discussion doesn't explain why you have that run call here. Isn't the one in the `init` function enough?

The problem is that (at least meanwhile) the init function fails due to some algs not available and then the test 'succeeds' simply because the test cases are not run and the test setup is not well designed to fail in such a case:
```
HARNESS_VERBOSE_FAILURE=yes make -j4 test TESTS=test_store V=1

make depend && make _tests
make[1]: Entering directory '/home/david/Siemens/proj/PKI/openssl/prepare1'
make[1]: Leaving directory '/home/david/Siemens/proj/PKI/openssl/prepare1'
make[1]: Entering directory '/home/david/Siemens/proj/PKI/openssl/prepare1'
( SRCTOP=. \
  BLDTOP=. \
  PERL="/usr/bin/perl" \
  EXE_EXT= \
  /usr/bin/perl ./test/run_tests.pl test_store )
90-test_store.t .. 
# The results of this test will end up in test-runs/test_store
Error, DSA key generation failed
../../util/wrap.pl ../../apps/openssl dsaparam -genkey 1024 => 1
# -----BEGIN EC PARAMETERS-----
# BggqhkjOPQMBBw==
# -----END EC PARAMETERS-----
# -----BEGIN EC PRIVATE KEY-----
# MHcCAQEEIPjXwljI5SKHSjCu7lAKtqfzEBHg3aUlOtBNGDyQ0B8yoAoGCCqGSM49
# AwEHoUQDQgAECpWVBwFOV0OD2/aXIfJ7KhruuYA4LLYCVMnEQXv6w/lcNepWvyYj
# Ae4jy5j8llmlZLzbQkIrRTR9ILllAkOlRA==
# -----END EC PRIVATE KEY-----
../../util/wrap.pl ../../apps/openssl ecparam -genkey -name prime256v1 => 0
1..177
../../../util/wrap.pl ../../../apps/openssl genrsa -out rsa-key-pkcs1.pem 2432 => 0
writing RSA key
../../../util/wrap.pl ../../../apps/openssl rsa -passout 'pass:password' -aes128 -in rsa-key-pkcs1.pem -out rsa-key-pkcs1-aes128.pem => 0
../../../util/wrap.pl ../../../apps/openssl ecparam -genkey -name prime256v1 -out ec-key-pkcs1.pem => 0
read EC key
writing EC key
../../../util/wrap.pl ../../../apps/openssl ec -passout 'pass:password' -aes128 -in ec-key-pkcs1.pem -out ec-key-pkcs1-aes128.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs8 -topk8 -nocrypt -in rsa-key-pkcs1.pem -out rsa-key-pkcs8.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs8 -topk8 -nocrypt -in ec-key-pkcs1.pem -out ec-key-pkcs8.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs8 -topk8 -passout 'pass:password' -v1 pbeWithSHA1And3-KeyTripleDES-CBC -in rsa-key-pkcs8.pem -out rsa-key-pkcs8-pbes1-sha1-3des.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs8 -topk8 -passout 'pass:password' -v1 pbeWithSHA1And3-KeyTripleDES-CBC -in rsa-key-pkcs8.pem -out rsa-key-pkcs8-pbes1-md5-des.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs8 -topk8 -passout 'pass:password' -v2 aes256 -v2prf hmacWithSHA1 -in rsa-key-pkcs8.pem -out rsa-key-pkcs8-pbes2-sha1.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs8 -topk8 -passout 'pass:password' -v2 aes256 -v2prf hmacWithSHA1 -in ec-key-pkcs8.pem -out ec-key-pkcs8-pbes2-sha1.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs8 -topk8 -passout 'pass:password' -v2 aes256 -v2prf hmacWithSHA256 -in rsa-key-pkcs8.pem -out rsa-key-pkcs8-pbes2-sha256.pem => 0
Generating a RSA private key
.......................................................................................................++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
..............++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
writing new private key to 'cakey.pem'
-----
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Common Name []:CA
../../../util/wrap.pl ../../../apps/openssl req -x509 -config ../../../test/recipes/90-test_store_data/ca.cnf -nodes -out cacert.pem -keyout cakey.pem => 0
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Common Name []:A user
User ID []:test
../../../util/wrap.pl ../../../apps/openssl req -new -config ../../../test/recipes/90-test_store_data/user.cnf -key rsa-key-pkcs8.pem -out rsa-cert.csr => 0
Request self-signature ok
subject=CN = A user, UID = test
Getting CA Private Key
../../../util/wrap.pl ../../../apps/openssl x509 -days 3650 -CA cacert.pem -CAkey cakey.pem -set_serial 1591182361 -req -in rsa-cert.csr -out rsa-cert.pem => 0
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Common Name []:A user
User ID []:test
../../../util/wrap.pl ../../../apps/openssl req -new -config ../../../test/recipes/90-test_store_data/user.cnf -key ec-key-pkcs8.pem -out ec-cert.csr => 0
Request self-signature ok
subject=CN = A user, UID = test
Getting CA Private Key
../../../util/wrap.pl ../../../apps/openssl x509 -days 3650 -CA cacert.pem -CAkey cakey.pem -set_serial 1591182361 -req -in ec-cert.csr -out ec-cert.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs12 -inkey rsa-key-pkcs8.pem -in rsa-cert.pem -passout 'pass:password' -chain -CAfile cacert.pem -export -macalg SHA1 -certpbe pbeWithSHA1And3-KeyTripleDES-CBC -keypbe pbeWithSHA1And3-KeyTripleDES-CBC -out rsa-key-sha1-3des-sha1.p12 => 0
../../../util/wrap.pl ../../../apps/openssl pkcs12 -inkey rsa-key-pkcs8.pem -in rsa-cert.pem -passout 'pass:password' -chain -CAfile cacert.pem -export -macalg SHA256 -certpbe pbeWithSHA1And3-KeyTripleDES-CBC -keypbe pbeWithSHA1And3-KeyTripleDES-CBC -out rsa-key-sha1-3des-sha256.p12 => 0
../../../util/wrap.pl ../../../apps/openssl pkcs12 -inkey rsa-key-pkcs8.pem -in rsa-cert.pem -passout 'pass:password' -chain -CAfile cacert.pem -export -macalg SHA256 -certpbe AES-256-CBC -keypbe AES-256-CBC -out rsa-key-aes256-cbc-sha256.p12 => 0
C087AAB8427F0000:error::digital envelope routines:(unknown function):fetch failed:crypto/evp/evp_fetch.c:307:Default library context, Algorithm (DES-CBC), Properties ()
C087AAB8427F0000:error::digital envelope routines:(unknown function):initialization error:crypto/evp/evp_enc.c:176:
C087AAB8427F0000:error::digital envelope routines:(unknown function):keygen failure:crypto/evp/evp_pbe.c:132:
C087AAB8427F0000:error::PKCS12 routines:(unknown function):pkcs12 algor cipherinit error:crypto/pkcs12/p12_decr.c:36:
C087AAB8427F0000:error::PKCS12 routines:(unknown function):encrypt error:crypto/pkcs12/p12_decr.c:133:
C087AAB8427F0000:error::PKCS12 routines:(unknown function):encrypt error:crypto/pkcs12/p12_add.c:119:
../../../util/wrap.pl ../../../apps/openssl pkcs12 -inkey rsa-key-pkcs8.pem -in rsa-cert.pem -passout 'pass:password' -chain -CAfile cacert.pem -export -macalg SHA1 -certpbe pbeWithMD5AndDES-CBC -keypbe pbeWithMD5AndDES-CBC -out rsa-key-md5-des-sha1.p12 => 1
ok 1 # skip failed initialisation
ok 2 # skip failed initialisation
ok 3 # skip failed initialisation
ok 4 # skip failed initialisation
ok 5 # skip failed initialisation
ok 6 # skip failed initialisation
ok 7 # skip failed initialisation
ok 8 # skip failed initialisation
ok 9 # skip failed initialisation
ok 10 # skip failed initialisation
ok 11 # skip failed initialisation
ok 12 # skip failed initialisation
ok 13 # skip failed initialisation
ok 14 # skip failed initialisation
ok 15 # skip failed initialisation
ok 16 # skip failed initialisation
ok 17 # skip failed initialisation
ok 18 # skip failed initialisation
ok 19 # skip failed initialisation
ok 20 # skip failed initialisation
ok 21 # skip failed initialisation
ok 22 # skip failed initialisation
ok 23 # skip failed initialisation
ok 24 # skip failed initialisation
ok 25 # skip failed initialisation
ok 26 # skip failed initialisation
ok 27 # skip failed initialisation
ok 28 # skip failed initialisation
ok 29 # skip failed initialisation
ok 30 # skip failed initialisation
ok 31 # skip failed initialisation
ok 32 # skip failed initialisation
ok 33 # skip failed initialisation
ok 34 # skip failed initialisation
ok 35 # skip failed initialisation
ok 36 # skip failed initialisation
ok 37 # skip failed initialisation
ok 38 # skip failed initialisation
ok 39 # skip failed initialisation
ok 40 # skip failed initialisation
ok 41 # skip failed initialisation
ok 42 # skip failed initialisation
ok 43 # skip failed initialisation
ok 44 # skip failed initialisation
ok 45 # skip failed initialisation
ok 46 # skip failed initialisation
ok 47 # skip failed initialisation
ok 48 # skip failed initialisation
ok 49 # skip failed initialisation
ok 50 # skip failed initialisation
ok 51 # skip failed initialisation
ok 52 # skip failed initialisation
ok 53 # skip failed initialisation
ok 54 # skip failed initialisation
ok 55 # skip failed initialisation
ok 56 # skip failed initialisation
ok 57 # skip failed initialisation
ok 58 # skip failed initialisation
ok 59 # skip failed initialisation
ok 60 # skip failed initialisation
ok 61 # skip failed initialisation
ok 62 # skip failed initialisation
ok 63 # skip failed initialisation
ok 64 # skip failed initialisation
ok 65 # skip failed initialisation
ok 66 # skip failed initialisation
ok 67 # skip failed initialisation
ok 68 # skip failed initialisation
ok 69 # skip failed initialisation
ok 70 # skip failed initialisation
ok 71 # skip failed initialisation
ok 72 # skip failed initialisation
ok 73 # skip failed initialisation
ok 74 # skip failed initialisation
ok 75 # skip failed initialisation
ok 76 # skip failed initialisation
ok 77 # skip failed initialisation
ok 78 # skip failed initialisation
ok 79 # skip failed initialisation
ok 80 # skip failed initialisation
ok 81 # skip failed initialisation
ok 82 # skip failed initialisation
ok 83 # skip failed initialisation
ok 84 # skip failed initialisation
ok 85 # skip failed initialisation
ok 86 # skip failed initialisation
ok 87 # skip failed initialisation
ok 88 # skip failed initialisation
ok 89 # skip failed initialisation
ok 90 # skip failed initialisation
ok 91 # skip failed initialisation
ok 92 # skip failed initialisation
ok 93 # skip failed initialisation
ok 94 # skip failed initialisation
ok 95 # skip failed initialisation
ok 96 # skip failed initialisation
ok 97 # skip failed initialisation
ok 98 # skip failed initialisation
ok 99 # skip failed initialisation
ok 100 # skip failed initialisation
ok 101 # skip failed initialisation
ok 102 # skip failed initialisation
ok 103 # skip failed initialisation
ok 104 # skip failed initialisation
ok 105 # skip failed initialisation
ok 106 # skip failed initialisation
ok 107 # skip failed initialisation
ok 108 # skip failed initialisation
ok 109 # skip failed initialisation
ok 110 # skip failed initialisation
ok 111 # skip failed initialisation
ok 112 # skip failed initialisation
ok 113 # skip failed initialisation
ok 114 # skip failed initialisation
ok 115 # skip failed initialisation
ok 116 # skip failed initialisation
ok 117 # skip failed initialisation
ok 118 # skip failed initialisation
ok 119 # skip failed initialisation
ok 120 # skip failed initialisation
ok 121 # skip failed initialisation
ok 122 # skip failed initialisation
ok 123 # skip failed initialisation
ok 124 # skip failed initialisation
ok 125 # skip failed initialisation
ok 126 # skip failed initialisation
ok 127 # skip failed initialisation
ok 128 # skip failed initialisation
ok 129 # skip failed initialisation
ok 130 # skip failed initialisation
ok 131 # skip failed initialisation
ok 132 # skip failed initialisation
ok 133 # skip failed initialisation
ok 134 # skip failed initialisation
ok 135 # skip failed initialisation
ok 136 # skip failed initialisation
ok 137 # skip failed initialisation
ok 138 # skip failed initialisation
ok 139 # skip failed initialisation
ok 140 # skip failed initialisation
ok 141 # skip failed initialisation
ok 142 # skip failed initialisation
ok 143 # skip failed initialisation
ok 144 # skip failed initialisation
ok 145 # skip failed initialisation
ok 146 # skip failed initialisation
ok 147 # skip failed initialisation
ok 148 # skip failed initialisation
ok 149 # skip failed initialisation
ok 150 # skip failed initialisation
ok 151 # skip failed initialisation
ok 152 # skip failed initialisation
ok 153 # skip failed initialisation
ok 154 # skip failed initialisation
ok 155 # skip failed initialisation
ok 156 # skip failed initialisation
ok 157 # skip failed initialisation
ok 158 # skip failed initialisation
ok 159 # skip failed initialisation
ok 160 # skip failed initialisation
ok 161 # skip failed initialisation
ok 162 # skip failed initialisation
ok 163 # skip failed initialisation
ok 164 # skip failed initialisation
ok 165 # skip failed initialisation
ok 166 # skip failed initialisation
ok 167 # skip failed initialisation
ok 168 # skip failed initialisation
ok 169 # skip failed initialisation
ok 170 # skip failed initialisation
ok 171 # skip failed initialisation
ok 172 # skip failed initialisation
ok 173 # skip failed initialisation
ok 174 # skip failed initialisation
ok 175 # skip failed initialisation
ok 176 # skip failed initialisation
ok 177 # skip failed initialisation
ok
All tests successful.
Files=1, Tests=177,  2 wallclock secs ( 0.03 usr  0.00 sys +  1.31 cusr  0.22 csys =  1.56 CPU)
Result: PASS
```
```

levitte replied:
> The problem is that (at least meanwhile) the init function fails due to some algs not available and then the test 'succeeds' simply because the test cases are not run and the test setup is not well designed to fail in such a case:

So there are two part in this:
1.  detecting that a feature is disabled.
2.  detecting that usage of a feature works.

The first part, I totally agree should affect what files to expect, i.e. variables such as `$use_dsa = !disabled("dsa")` is exactly correct.
The second part should remain in `init`.  If it doesn't work, then the variable `$use_dsa` hasn't been checked properly, or one particular command invocation has failed and needs to be corrected.

The specific failure to init with pbeWithMD5AndDES-CBC can, I think, be solved like this:

``` diff
diff --git a/test/recipes/90-test_store.t b/test/recipes/90-test_store.t
index 3e2e69f439..3a58a2fd94 100644
--- a/test/recipes/90-test_store.t
+++ b/test/recipes/90-test_store.t
@@ -17,6 +17,8 @@ setup($test_name);
 
 my $mingw = config('target') =~ m|^mingw|;
 
+my $use_des = !disabled("des");
+
 my @noexist_files =
     ( "test/blahdiblah.pem",
       "test/blahdibleh.der" );
@@ -336,7 +338,9 @@ sub init {
                           my %pbes =
                               (
                                "sha1-3des" => "pbeWithSHA1And3-KeyTripleDES-CBC",
-                               "md5-des" => "pbeWithMD5AndDES-CBC",
+                               ($use_des
+                                ? ( "md5-des" => "pbeWithMD5AndDES-CBC" )
+                                : () ),
                                "aes256-cbc" => "AES-256-CBC",
                               );
                           my %macalgs =
```

levitte replied:
A trick to know with perl is that lists are flattened, i.e. `(1, (2, 3), 4)` becomes `(1, 2, 3, 4)`, and that `=> ` is syntax sugar for a list of a pair, i.e. `a => 1` is the same as `('a', 1)`

DDvO replied:
Concerning the `openssl dsaparam` call used in the test init, I meanwhile found out that the key size of 1024 is no longer acceptable by the default and after changing this to an acceptable value such as 2048, dsaparam crashes. See the new bug report #12047.

As a workaround I resorted to using an already existing param file, 
which I found in `test/recipes/04-test_pem_data/dsaparam.pem`.

DDvO replied:
Meanwhile I realized that the tests anyway should better not generate fresh RSA and DSA keys since this is just a waste of time/computations and the focus of these test is just reading various credential files via the OSSL_STORE API.
So I've changed this to using pre-computed RSA keys and pre-computed DSA parameters.

mattcaswell commented on test/recipes/90-test_store.t:
```diff
@@ -223,24 +234,28 @@ sub init {
             && run(app(["openssl", "dsa", "-passout", "pass:password", "-aes128",
                         "-in", "dsa-key-pkcs1.pem",
                         "-out", "dsa-key-pkcs1-aes128.pem"]))
+                      }, grep(/dsa-key-pkcs1\.pem$/, @generated_files))
```
The indentation in this "sub" is quite confusing

mattcaswell replied:
I'm trying to figure out why we need to use `runall` here (and similar occurrences) below. AFAICT we only ever run this sub once?.....ahh.......`dsa-key-pkcs1.pem`, may not exist in `@generated_files` at all - right? That's quite a convoluted way of writing an `||` condition, i.e. `!$use_dsa || run(app(....))`

DDvO replied:
Good point; I've changed the guard accordingly.

DDvO commented:
I've meanwhile adapted `OSSL_STORE_close()` such that it succeeds gracefully when given a NULL argument, like any `free()` function does.

DDvO commented:
Note that I also fixed some issues in the OSSL_STORE documentation.

t8m commented on crypto/store/store_lib.c:
```diff
@@ -231,6 +231,8 @@ int OSSL_STORE_close(OSSL_STORE_CTX *ctx)
 {
     int loader_ret;
 
+    if (ctx == NULL)
```
This should be backported to 1.1.1 to avoid changing the semantics. So the callers can depend on the NULL check.

DDvO replied:
PR for backport this enhancement of (allowing `NULL` arg for `OSSL_STORE_close()`) is in #11912.

levitte commented:
Regarding `OSSL_STORE_close()` being a kind of `free()`, it's semantically closer to `fclose()`. What's the result of `fclose(NULL)`? :wink:

DDvO commented:
> Regarding `OSSL_STORE_close()` being a kind of `free()`, it's semantically closer to `fclose()`. What's the result of `fclose(NULL)`? wink

I know that the semantics of `fclose(NULL)` is officially undefined,
and I believe that this is purely for historical reasons,
IMHO it is preferrable to define the behavior in this case as no-op. 
For `OSSL_STORE_close()` we have the chance to do better :smile:

DDvO commented:
I've just done the same enhancements of `OSSL_STORE_open.pod` as in #11912.

levitte commented:
For commit messages, I disagree with "Allow NULL argument (as with *_free()) to OSSL_STORE_close()".  Please make it "Allow NULL argument to OSSL_STORE_close()".
My reasoning is that *although* `OSSL_STORE_close()` does free its argument, it's semantically closer to `fclose()`, the freeing part is just an implementation detail.

t8m commented on crypto/store/loader_file.c:
```diff
@@ -256,53 +255,57 @@ static OSSL_STORE_INFO *try_decode_PKCS12(const char *pem_name,
+                if ((ctx = sk_OSSL_STORE_INFO_new_null()) != NULL) {
+                    if (pkey != NULL) {
+                        if ((osi_pkey = OSSL_STORE_INFO_new_PKEY(pkey)) != NULL
+                            && (pkey = NULL) == NULL
```
This construct is 'interesting'. On the other hand the intention is clear, so I just hope that Coverity won't barf on it.

DDvO replied:
> This construct is 'interesting'. On the other hand the intention is clear, so I just hope that Coverity won't barf on it.

Yeah, I'm proud about this nice trick avoiding lengithy case distinctions :smile:

richsalz replied:
It's tricky code, and isn't something to be proud about.  It should be rewritten.  As that is unlikely to happen :(, please comment it.

DDvO replied:
All right, added an explaining comment in both instances.

t8m commented:
The commit message can be amended without a re-review or even during the merge.

DDvO commented:
> The commit message can be amended without a re-review or even during the merge.

Thanks @t8m for helping out toward finalizing this PR!
I've just rebased it, squashed two related commits (regarding doc of `OSSL_STORE`),
and re-worded one commit as requested by @levitte.

Unless there is an objection I'm going to merge this tomorrow.

t8m commented:
Still approved after the rebase, for the record.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

levitte commented on crypto/store/loader_file.c:
```diff
@@ -256,53 +255,59 @@ static OSSL_STORE_INFO *try_decode_PKCS12(const char *pem_name,
+                int ok = 1;
+
+                if ((ctx = sk_OSSL_STORE_INFO_new_null()) != NULL) {
+                    if (pkey != NULL) {
```
According to documentation, if `PKCS12_parse()` is successful, `pkey` can't be NULL.  If it is at this point, then there's a bug in `PKCS12_parse()` that should be fixed.

levitte replied:
Ah, never mind...  the PKCS#12 blob might not have a key, gotcha...

levitte commented on doc/man3/OSSL_STORE_INFO.pod:
```diff
@@ -54,25 +54,15 @@ loaders to create B<OSSL_STORE_INFO> holders.
-STORE_INFO_get0_TYPE().
-The life time of this extracted object is as long as the life time of
-the B<OSSL_STORE_INFO> it was extracted from, so care should be taken not
-to free the latter too early.
```
The warning about lifetime should be retained.  Since you removed it here, maybe it can be added back with the description of the get0 functions (feel free to edit it to be suitable)

DDvO replied:
I removed the section about `STORE_INFO_get0_TYPE` and `-STORE_INFO_get1_TYPE` because they do not (any more) exist. Yet I just realized that likely `STORE_INFO_get_type` can be used instead, so I restored the first sentence of that hint, now referring to that function.

Still I see no point in restoring 
```
The life time of this extracted object is as long as the life time of
the B<OSSL_STORE_INFO> it was extracted from, so care should be taken not
to free the latter too early.
```
because its result is not an allocated object but a simple number.

levitte replied:
> I removed the section about STORE_INFO_get0_TYPE and -STORE_INFO_get1_TYPE because they do not (any more) exist.

What on earth are you talking about?
... OH, I see, you took the upper case `TYPE` literally, instead of just being a placeholder word for the different types of info (`NAME`, `PARAMS`, `PKEY`, `CERT`, `CRL`).

``` console
$ grep OSSL_STORE_INFO_get include/openssl/store.h
 * extracted with OSSL_STORE_INFO_get0_PKEY(), OSSL_STORE_INFO_get0_CERT(), ...
int OSSL_STORE_INFO_get_type(const OSSL_STORE_INFO *info);
const char *OSSL_STORE_INFO_get0_NAME(const OSSL_STORE_INFO *info);
char *OSSL_STORE_INFO_get1_NAME(const OSSL_STORE_INFO *info);
const char *OSSL_STORE_INFO_get0_NAME_description(const OSSL_STORE_INFO *info);
char *OSSL_STORE_INFO_get1_NAME_description(const OSSL_STORE_INFO *info);
EVP_PKEY *OSSL_STORE_INFO_get0_PARAMS(const OSSL_STORE_INFO *info);
EVP_PKEY *OSSL_STORE_INFO_get1_PARAMS(const OSSL_STORE_INFO *info);
EVP_PKEY *OSSL_STORE_INFO_get0_PKEY(const OSSL_STORE_INFO *info);
EVP_PKEY *OSSL_STORE_INFO_get1_PKEY(const OSSL_STORE_INFO *info);
X509 *OSSL_STORE_INFO_get0_CERT(const OSSL_STORE_INFO *info);
X509 *OSSL_STORE_INFO_get1_CERT(const OSSL_STORE_INFO *info);
X509_CRL *OSSL_STORE_INFO_get0_CRL(const OSSL_STORE_INFO *info);
X509_CRL *OSSL_STORE_INFO_get1_CRL(const OSSL_STORE_INFO *info);
```

DDvO replied:
Argh!! So this was just a misunderstanding.
I've restored the original test, but using `<TYPE>` instead of `TYPE` to prevent that confusion and adding the list of potential `TYPE` instances.

levitte commented:
> When loading PKCS#12 files that include a cert/ca chain the chain container was not deallocated.

Looking back, I'm surprised we haven't seen that already.  Is there a way to demonstrate this with existing master, or even 1.1.1?

DDvO commented:
> > When loading PKCS#12 files that include a cert/ca chain the chain container was not deallocated.
> 
> Looking back, I'm surprised we haven't seen that already. Is there a way to demonstrate this with existing master, or even 1.1.1?

Oops, I put my answer further up in the discussion; see there.

levitte commented:
> And so far the tests did not include a chain in the PKCS#12 file(s) parsed.

Thanks, that's what I suspected.  So we need one to test the mem leak, can one be provided, or musst we count on one being generated?

DDvO commented:
I had already added a chain to some of the PKCS#12 test cases:
```
+++ b/test/recipes/90-test_store.t
@@ -374,6 +374,7 @@ sub init {
                           }
                           run(app(["openssl", "pkcs12", "-inkey", $srckey,
                                    "-in", $srccert, "-passout", "pass:password",
+                                   "-chain", "-CAfile", "cacert.pem",
                                    "-export", "-macalg", $macalg,
                                    "-certpbe", $certpbe, "-keypbe", $keypbe,
                                    "-out", $dstfile]));
```
These would fail if if the mem leak still existed - provided not all the test were disabled due to init fail issue discussed above.

DDvO commented:
Meanwhile I've added also `$use_des` and successfully tested various combinations of
negating `$use_md5`, `$use_des`, `$use_dsa`, and `$use_ecc`.

DDvO commented:
I also found that, due to the so far 'successful' skip of the tests in case their init failed,
the test cert generation using the rather new CA configuration file `test/ca-and-certs.cnf`
went wrong without being noticed by its author (apparently @richsalz in commit 4e6e57cfcdd75b827ff7171927d87e95b5b86ae8).
I've corrected this as well and made sure that the tests fail if their init fails,
to prevent such obscure failure issues for the future.

DDvO commented:
@levitte wrote: 

> Will need re-review

Yes. Now the PR is ready for that.

t8m commented:
For some reason Travis build is not getting executed.

DDvO commented:
> For some reason Travis build is not getting executed.

I kicked it by closing and re-opening the PR, but this did not help.
Apparently a more general issue?

DDvO commented:
It looks like Travis just had a delay starting up. It's running again.

levitte commented on test/recipes/90-test_store.t:
```diff
@@ -85,7 +95,7 @@ plan tests => $n;
  SKIP:
     {
-        skip "failed initialisation", $n unless init();
+        init() or die "init failed";
```
s/`die "init failed"`/`BAIL_OUT("init failed")`/

See `man Test::More`

DDvO replied:
Done, also for the two other `die` instances.

DDvO replied:
Hmm, meanwhile I've notices how Travis behaves on `BAIL_OUT()`- see for instance https://travis-ci.org/github/openssl/openssl/jobs/695192059:
```
90-test_sslapi.t ................... ok
90-test_sslbuffers.t ............... ok
Bailout called.  Further testing stopped:  init failed
FAILED--Further testing stopped: init failed

Makefile:2013: recipe for target '_tests' failed
make: *** [_tests] Error 255
/home/travis/.travis/functions: line 607: 16180 Terminated              travis_jigger "${!}" "${timeout}" "${cmd[@]}"
** FAILED -- MAKE TEST
```

On `BAIL_OUT()` all the tests end abruptly, and without even stating for which test the failure occurred. On the other hand, using `die` only the current test is terminated and the log shows which test failed, and then the rest of the test continues.

For these reasons I've reverted `s/die "init failed"/BAIL_OUT("init failed")/` (as well as the other two such changes in that script.

levitte replied:
Oh!  You're right, I had forgotten that it's that brutal.  Ok then!

DDvO replied:
Fine. Can you re-approve then?

DDvO commented:
There was Travis failure, which pointed out that PKCS#12 generation is not possible in case no-des.
Fixup committed.

DDvO commented:
It turns out that there was another init failure in case `no-des`, which I've handled now.
@levitte, I meanwhile understand why when authoring `90-test_store.t` you took the shortcut of simply letting the tests `pass' (for void) if anything went wrong in `init()`.

I'm confident that now all relevant algorithm exclusion cases are finally sorted out and ask you to re-confirm your approval.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

DDvO commented:
Thanks @levitte for the swift re-review and re-confirmation.
I've just rebased and squashed the fixups (to be merged on Monday morning).

DDvO commented:
Merged - thanks!
